### PR TITLE
Ensure global defers run on exit

### PR DESCRIPTION
**Pull request by hadley**
Opening this primarily for discussion — this change means that when `withr::defer()` is called interactively (i.e. from the global environment), it will be run when you end the session.

I'm not 100% sure it's a good idea (or if it's been discussed and rejected before) but this is how I expected it to wrk. I started looking into because I noticed that my tests weren't always running cleanup code, and that lead to a simple experiment confirming that `withr::defer(print("Hi"))` is not executed when you quit R.

cc @jennybc, @jimhester, @lionel-

### CODE REVIEW

**Comment by jennybc** [comment]
We have discussed this before, on the original PR, started [around here](https://github.com/r-lib/withr/pull/76#pullrequestreview-394573897).

**Comment by jennybc** [comment]
In terms of the workflow aspect, I have something in my `.Rprofile` that makes `dr` a shortcut for `withr::deferred_run()`, which I use heavily when I'm developing, e.g. tests, that use `defer()`.

**Comment by jennybc** [comment]
See also #136

**Comment by hadley** [comment]
I now feel more strongly that the non-finalizer approach is suboptimal because it relies on me to remember that I have to run `deferred_run()`, which is very hard.

**Comment by jennybc** [comment]
I don't have a strong opinion about this. I guess I prefer the *status quo* and I am certainly very used to it, but I can also make peace with the proposed change.

**Comment by lionel-** [comment]
The session finalizer comes in addition to the existing features, so I guess it makes sense. I wouldn't rely on it being run for important things though.

In general I don't have any strong feeling about global `defer()` except that we need to fix #136 so that the message isn't displayed in non-top-level global evaluations which are actually cleaned up automatically (I'm still happy to send a PR).

**Comment by hadley** [comment]
@jimhester if you want a test, I think I'd have to add a dependency on callr. I can't see any other way except to start another R process.

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # withr (development version)
 
+* Handlers registered with the global environment (as happens when `local_()` 
+  is run at the top-level, outside a function) are now automatically run
+  when the R session ends.
```
Comment (jimhester) [inline NEWS.md:5]: Could you add the PR number here?

**Comment by jimhester** [comment]
I think it is ok to skip the test here.

**Comment by hadley** [comment]
Will need an update after #176

**Review by jimhester — APPROVED** [review]

**Pull request merged**